### ISSUE TITLE
Pull-based Model

### ISSUE THREAD

**Issue by RichiCoder1** [argoproj/argo-cd#4512]
# Summary

Enable a daemon-based pull model where a cluster agent manages diffs and reconciliation, and applies changes rather than the central Argo App.

# Motivation

With the launch of [Rancher's Fleet](https://rancher.com/docs/rancher/v2.x/en/deploy-across-clusters/fleet/) and [PipeCD](https://pipecd.dev/blog/2020/10/06/announcing-pipecd/), as well as the general friend-of-Argo Flux, there's a big interest in enabling a pull model while still having a central control and observability plan.

Wanted to start a discussion of whether this model could and should be implemented for ArgoCD

# Proposal

No concrete proposal, but specifically only proposing shifting the portion of Argo that interacts with the clusters to the clusters themselves. The daemon would then register with ArgoCD rather than ArgoCD managing via the K8s API>

**Comment by jessesuen** [comment]
With the new sharding feature coming in v1.8, the groundwork for this is starting to become possible. But, there would be a ton of extra effort surrounding the agent-controlplane communication, update and maintenance considrations.

**Comment by RichiCoder1** [comment]
Absolutely, and may be out of scope or too large and effort. Was curious about starting a discussion though.

**Comment by blakepettersson** [comment]
Something like this is in the works: #11879
